keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,191 @@
+3.4 to efficiently determine how up-to-date kubernetes watch caches are then
+serve reads from the watch cache when they are sufficiently up-to-date.
+
+## Motivation
+
+Intuitively, we expect that serving reads from the watch cache to be more
+performant and scalable than reading them from etcd, deserializing them,
+converting them to the desired type and then garbage collecting all the
+objects that were allocated during the read. 
+
+We will need to measure the impact to performance and scalability, but we
+believe there is a significant opportunity and would like to explore it.
+
+In addition to the general scale and performance opportunity, we also aim
+to resolve a specific problem. The long standing "stale read" issue
+(https://github.com/kubernetes/kubernetes/issues/59848) is still open because
+reflectors default to resourceVersion=”0” for their initial list requests. If
+the reflectors instead use a consistent read for their initial list request,
+they could not "going back in time" when components are restarted, which can
+curently happen if the initial list request is served from a stale watch cache
+with data much older than the reflector has previously observed or if the
+api-server or etcd are partitioned.
+
+We have held off on switching reflectors to using consistent read for the
+initial list, even though we know it is more correct, due to concerns with the
+impact on large scale use cases. But if we serve consistent reads from cache,
+there would be very little difference in scalability to how the kube-apiserver
+serves the resourceVersion="0" list requests from reflectors today.
+
+### Goals
+
+- Serve consistent reads from the watch cache for etcd 3.4 or newer
+- Resolve the "stale read" problem (https://github.com/kubernetes/kubernetes/issues/59848)
+- Improve the scailability and performance of Kubernetes for Get and List requests, when the watch cache is enabled
+
+### Non-Goals
+
+- Serve consistent reads from the watch cache for etcd 3.3 or earlier
+
+## Proposal
+
+Allow the etcd minor version to be specified in the kube-apiserver `--storage-backend` flag, e.g. `--storage-backend=etcd3.4`.  For backward compatibility, if `--storage-backend=etcd3` is provided, it will treat this as `etcd3.0`.
+
+Since etcd progress notify is only available in etcd 3.4. We would only serve consistent reads from cache if `--storage-backend` is set to `etcd3.4` or higher.
+For etcd 3.3-, we would continue to serve consistent reads from etcd.
+
+Guard this optmization behind `WatchCacheConsistentReads` feature gate, and only server consistent reads from cache if it is enabled.
+
+When an consistent read request is received and the watch cache is enabled:
+- Use a range(limit=0) request to get the current revision from etcd for the resource type being served. Etcd can serve this type of request efficiently, and the resulting revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
```

Comment:
Maybe I'm completely wrong, but let me throw this idea:
- IIUC, the requests to etcd with limit=N set, result in forking the underlying tree to allow serving the "continuation" request later; so intuitively, if we will be sending a lot of those, that may be relatively expensive
- what if instead, we will be sending requests with `WithCountOnly` set? is computing count much more expensive or is it also computed in constant time (or close to constant?)

@jpbetz @jingyih

Reply:
> IIUC, the requests to etcd with limit=N set, result in forking the underlying tree to allow serving the "continuation" request later; so intuitively, if we will be sending a lot of those, that may be relatively expensive

When limit=N set, etcd server reads N+1 keys from disk. 
@jpbetz limit=0 means no limit, so it is not what we want.

> what if instead, we will be sending requests with WithCountOnly set? 

This is a good idea. WithCountOnly is served from the in memory index tree in etcd - no need to read from disk.

We could do some experiments to measure which is the cheapest way to get the current revision from etcd.

Reply:
> IIUC, the requests to etcd with limit=N set, result in forking the underlying tree to allow serving the "continuation" request later;

BTW, "continuation" is achieved by asking for the same revision in the next request.